int64 best_dir
float64[] directions